src/main/java/com/microsoftopentechnologies/windowsazurestorage/service/DownloadService.java

```diff
@@ -68,29 +70,45 @@ protected int downloadBlobs(final Iterable<ListBlobItem> blobItems)
         return filesDownloaded;
     }
 
-    protected void downloadBlob(final CloudBlob blob)
-            throws WAStorageException {
+    protected void downloadSingleFile(final CloudFile cloudFile) throws WAStorageException {
         final DownloadServiceData serviceData = getServiceData();
         try {
-            final FilePath downloadDir = serviceData.getDownloadDir();
-            FilePath downloadFile = new FilePath(downloadDir, blob.getName());
-
-            // That filepath will contain all the directories and explicit virtual
-            // paths, so if the user wanted it flattened, grab just the file name and
-            // recreate the file path
-            if (serviceData.isFlattenDirectories()) {
-                downloadFile = new FilePath(downloadDir, downloadFile.getName());
+            println("Downloading file:" + cloudFile.getUri().toString());
+            final FilePath destFile = destinationFilePath(cloudFile.getName());
+
+            final long startTime = System.currentTimeMillis();
+            try (OutputStream fos = destFile.write()) {
+                cloudFile.download(fos, null, new FileRequestOptions(), Utils.updateUserAgent());
             }
+            final long endTime = System.currentTimeMillis();
+            println(String.format(
+                    "blob %s is downloaded to %s in %s",
+                    cloudFile.getName(), destFile.getParent(), getTime(endTime - startTime)));
+
+            if (serviceData.isDeleteFromAzureAfterDownload()) {
+                cloudFile.deleteIfExists();
+                println("cloud file " + cloudFile.getName() + " is deleted from Azure.");
+            }
+        } catch (Exception e) {
+            throw new WAStorageException(e.getMessage(), e.getCause());
```

Comment:
We may lose one frame of the stack trace if we use `e.getCause()`, should be `e` here.